target_postgres/__init__.py

```diff
@@ -297,47 +358,73 @@ def flush_streams(
-                # Create bookmark key if not exists
-                if 'bookmarks' not in flushed_state:
-                    flushed_state['bookmarks'] = {}
-                # Copy the stream bookmark from the latest state
-                flushed_state['bookmarks'][stream] = copy.deepcopy(state['bookmarks'][stream])
-
-        # If we flush every bucket use the latest state
-        else:
-            flushed_state = copy.deepcopy(state)
+        streams[stream].clear()
+    
+    if len(streams_to_flush) > 0:
+        for stream in streams_to_flush:
+            if stream not in flushed_state['bookmarks']:
+                flushed_state['bookmarks'][stream] = []
+            elif isinstance(flushed_state['bookmarks'][stream], dict):
+                flushed_state['bookmarks'][stream] = [copy.deepcopy(flushed_state['bookmarks'][stream])]
+            flushed_state['bookmarks'][stream].extend(batch_states['bookmarks'][stream])
+            for key in flushed_state['summary'][stream]:
+                flushed_state['summary'][stream][key] += batch_states['summary'][stream][key]
 
     # Return with state message with flushed positions
     return flushed_state
 
 
 # pylint: disable=too-many-arguments
-def load_stream_batch(stream, records_to_load, row_count, db_sync, delete_rows=False, temp_dir=None):
+def load_stream_batch(stream, records_to_load, row_count, db_sync:DbSync, delete_rows=False, temp_dir=None):
     """Load a batch of records and do post load operations, like creating
     or deleting rows"""
     # Load into Postgres
-    if row_count[stream] > 0:
-        flush_records(stream, records_to_load, row_count[stream], db_sync, temp_dir)
-
-    # Load finished, create indices if required
-    db_sync.create_indices(stream)
-
-    # Delete soft-deleted, flagged rows - where _sdc_deleted at is not null
-    if delete_rows:
-        db_sync.delete_rows(stream)
-
-    # reset row count for the current stream
-    row_count[stream] = 0
-
-
+    batch_state = {
+        "bookmarks": {},
+        "summary": {
+            "success": 0,
+            "fail": 0,
+            "existing": 0,
+            "updated": 0
+        }
+    }
+    try:
+        if row_count[stream] > 0:
+            batch_state = flush_records(stream, records_to_load, row_count[stream], db_sync, temp_dir)
+
+        # Load finished, create indices if required
+        db_sync.create_indices(stream)
+
+        # Delete soft-deleted, flagged rows - where _sdc_deleted at is not null
+        if delete_rows:
+            db_sync.delete_rows(stream)
+
+        # reset row count for the current stream
+        row_count[stream] = 0
+    except Exception as e:
+        batch_state["summary"][stream]["fail"] += len(records_to_load)
+        for record in records_to_load.values():
+            batch_state["bookmarks"][stream].append({
+                "hash": build_record_hash(record),
+                "success": False,
+                "external_id": record.get('cid',""),
```

Comment:
What is `cid` in this context? Is it guaranteed to always exist or was it just an example from your `data.singer`? In other targets we check if the user has supplied `externalId` and if not we don't include this `external_id` in the `bookmark` for this record

Reply:
It was particular to the data.singer. Fixed to check if externalId was supplied.